File: frontend/src/services/api/feedbackService.ts
import { apiClient } from './client';

export interface ApiResponse<T = any> {
  success: boolean;
  data?: T;
  message?: string;
  error?: string;
}

export interface SubmitFeedbackRequest {
  alertId: string;
  responderId?: number;
  rating: number;
  feedback?: string;
  summaryNote?: string;
  wasHelpful?: boolean;
  feedbackType?: string;
}

export interface UserRatingStats {
  user_id: number;
  average_rating: number;
  total_ratings: number;
  five_star_count: number;
  four_star_count: number;
  three_star_count: number;
  two_star_count: number;
  one_star_count: number;
  recent_feedback: Array<{
    id: number;
    alert_id: string;
    user_id: number;
    responder_id?: number;
    rating: number;
    feedback?: string;
    summary_note?: string;
    was_helpful?: boolean;
    feedback_type: string;
    created_at: string;
  }>;
}

export interface TopRatedUser {
  id: number;
  name: string;
  phone: string;
  average_rating: number;
  total_ratings: number;
}

export const feedbackService = {
  /**
   * Submit feedback for an alert
   */
  submitFeedback: async (data: SubmitFeedbackRequest): Promise<{ feedback_id: number }> => {
    const response = await apiClient.post<ApiResponse<{ feedback_id: number }>>('/feedback', {
      alert_id: data.alertId,
      responder_id: data.responderId,
      rating: data.rating,
      feedback: data.feedback,
      summary_note: data.summaryNote,
      was_helpful: data.wasHelpful,
      feedback_type: data.feedbackType || 'resolution',
    });

    const legacyFeedbackId = (response.data as ApiResponse<{ feedback_id: number }> & { feedback_id?: number }).feedback_id;

    if (!response.data.data && legacyFeedbackId) {
      return { feedback_id: legacyFeedbackId };
    }

    return response.data.data!;
  },

  /**
   * Get rating statistics for a user
   */
  getUserRatings: async (userId: number): Promise<UserRatingStats> => {
    const response = await apiClient.get<UserRatingStats>(`/users/${userId}/ratings`);
    return response.data;
  },

  /**
   * Get top rated users
   */
  getTopRatedUsers: async (limit: number = 10): Promise<TopRatedUser[]> => {
    const response = await apiClient.get<ApiResponse<{ users: TopRatedUser[]; count: number }>>(
      '/users/top-rated',
      { params: { limit } }
    );
    return response.data.data?.users || [];
  },
};

File: backend/internal/services/feedback_service.go
package services

import (
	"fmt"

	"github.com/obsidiannnn/Safe-Around/backend/internal/models"
	"gorm.io/gorm"
)

type FeedbackService struct {
	db *gorm.DB
}

func NewFeedbackService(db *gorm.DB) *FeedbackService {
	return &FeedbackService{db: db}
}

// SubmitFeedback creates a new feedback entry and updates user ratings
func (s *FeedbackService) SubmitFeedback(feedback *models.UserFeedback) error {
	tx := s.db.Begin()
	defer func() {
		if r := recover(); r != nil {
			tx.Rollback()
		}
	}()

	// Create the feedback record
	if err := tx.Create(feedback).Error; err != nil {
		tx.Rollback()
		return fmt.Errorf("failed to create feedback: %w", err)
	}

	// Update responder ratings if this is responder feedback
	if feedback.ResponderID != nil {
		if err := s.updateUserRatings(tx, *feedback.ResponderID); err != nil {
			tx.Rollback()
			return fmt.Errorf("failed to update user ratings: %w", err)
		}
	}

	return tx.Commit().Error
}

// updateUserRatings recalculates and updates user rating statistics
func (s *FeedbackService) updateUserRatings(tx *gorm.DB, userID uint) error {
	var stats struct {
		TotalRatings   int64
		AverageRating  float64
		FiveStarCount  int64
		FourStarCount  int64
		ThreeStarCount int64
		TwoStarCount   int64
		OneStarCount   int64
	}

	// Calculate rating statistics
	if err := tx.Model(&models.UserFeedback{}).
		Select(`
			COUNT(*) as total_ratings,
			AVG(rating::decimal) as average_rating,
			COUNT(CASE WHEN rating = 5 THEN 1 END) as five_star_count,
			COUNT(CASE WHEN rating = 4 THEN 1 END) as four_star_count,
			COUNT(CASE WHEN rating = 3 THEN 1 END) as three_star_count,
			COUNT(CASE WHEN rating = 2 THEN 1 END) as two_star_count,
			COUNT(CASE WHEN rating = 1 THEN 1 END) as one_star_count
		`).
		Where("responder_id = ? AND deleted_at IS NULL", userID).
		Scan(&stats).Error; err != nil {
		return fmt.Errorf("failed to calculate rating statistics: %w", err)
	}

	// Update user record with new statistics
	updates := map[string]interface{}{
		"total_ratings":     stats.TotalRatings,
		"average_rating":    stats.AverageRating,
		"five_star_count":   stats.FiveStarCount,
		"four_star_count":   stats.FourStarCount,
		"three_star_count":  stats.ThreeStarCount,
		"two_star_count":    stats.TwoStarCount,
		"one_star_count":    stats.OneStarCount,
	}

	if err := tx.Model(&models.User{}).Where("id = ?", userID).Updates(updates).Error; err != nil {
		return fmt.Errorf("failed to update user rating statistics: %w", err)
	}

	return nil
}

// GetUserRatingStats returns detailed rating statistics for a user
func (s *FeedbackService) GetUserRatingStats(userID uint) (*UserRatingStats, error) {
	var user models.User
	if err := s.db.First(&user, userID).Error; err != nil {
		return nil, fmt.Errorf("user not found: %w", err)
	}

	// Get recent feedback
	var recentFeedback []models.UserFeedback
	if err := s.db.Where("responder_id = ?", userID).
		Order("created_at DESC").
		Limit(10).
		Find(&recentFeedback).Error; err != nil {
		return nil, fmt.Errorf("failed to get recent feedback: %w", err)
	}

	return &UserRatingStats{
		UserID:         user.ID,
		AverageRating:  user.AverageRating,
		TotalRatings:   user.TotalRatings,
		FiveStarCount:  user.FiveStarCount,
		FourStarCount:  user.FourStarCount,
		ThreeStarCount: user.ThreeStarCount,
		TwoStarCount:   user.TwoStarCount,
		OneStarCount:   user.OneStarCount,
		RecentFeedback: recentFeedback,
	}, nil
}

// GetTopRatedUsers returns users with highest ratings
func (s *FeedbackService) GetTopRatedUsers(limit int) ([]models.User, error) {
	var users []models.User
	if err := s.db.Where("total_ratings >= ?", 5). // Minimum 5 ratings to be considered
		Order("average_rating DESC, total_ratings DESC").
		Limit(limit).
		Find(&users).Error; err != nil {
		return nil, fmt.Errorf("failed to get top rated users: %w", err)
	}

	return users, nil
}

type UserRatingStats struct {
	UserID         uint                    `json:"user_id"`
	AverageRating  float64                 `json:"average_rating"`
	TotalRatings   int                     `json:"total_ratings"`
	FiveStarCount  int                     `json:"five_star_count"`
	FourStarCount  int                     `json:"four_star_count"`
	ThreeStarCount int                     `json:"three_star_count"`
	TwoStarCount   int                     `json:"two_star_count"`
	OneStarCount   int                     `json:"one_star_count"`
	RecentFeedback []models.UserFeedback   `json:"recent_feedback"`
}